### ISSUE TITLE
Getting "access denied" when switching S3 backends in multiple AWS account setup

### ISSUE THREAD

**Issue by cgswong** [hashicorp/terraform#18007]
### Terraform Version
```
$ terraform --version
Terraform v0.11.7
+ provider.aws v1.17.0
```

### Terraform Configuration Files
```hcl
terraform.tf:
terraform {
  backend "s3" {}
}

backend-prod.tfvars:
bucket = "terraform-111111111111"

key = "r53cwlog/prod/terraform.tfstate"

region = "us-east-1"

encrypt = "true"

dynamodb_table = "r53cwlog_tf_state_lock"


backend-non-prod.tfvars:
bucket = "terraform-222222222222"

key = "r53cwlog/non-prod/terraform.tfstate"

region = "us-east-1"

encrypt = "true"

dynamodb_table = "r53cwlog_tf_state_lock"

```

### Debug Output

https://gist.github.com/cgswong/40ab841741137b3631c5bef02762f905

### Expected Behavior

Terraform switching backends from S3 in account A to account B successfully.

```bash
terraform init -backend-config=env/backend-prod.tfvars
```

### Actual Behavior

Terraform backend switch results in "Access Denied" error using account with AWS Administrator privileges:

```bash
$ terraform init -backend-config=env/backend-prod.tfvars

Initializing the backend...
Backend configuration changed!

Terraform has detected that the configuration specified for the backend
has changed. Terraform will now check for existing state in the backends.


Error inspecting states in the "s3" backend:
    AccessDenied: Access Denied
	status code: 403, request id: 809BFEAC6F189586, host id: oR6bxLZCwV/wB3u3HZXE1bXgeGnZTwJD7Z4EINGtrl1f4ea+qtUEFfiJT/Uwxc4CfGB3onyr4ow=

Prior to changing backends, Terraform inspects the source and destination
states to determine what kind of migration steps need to be taken, if any.
Terraform failed to load the states. The data in both the source and the
destination remain unmodified. Please resolve the above error and try again.
```

### Steps to Reproduce

1. With AWS environment variables set for account A run:

    ```bash
    terraform init -backend-config=env/backend-non-prod.tfvars
    terraform plan -var-file=env/non-prod.tfvars -out=non-prod.tfplan
    terraform apply non-prod.tfplan
    ```

2. With AWS environment variables set for account B run:

    ```bash
    terraform init -backend-config=env/backend-prod.tfvars
    ```

### Additional Context

Error occurs with and without Makefile. Confirmed role being used has administrator privileges, and can read/write to S3 bucket hosting state.

**Comment by jbardin** [comment]
Hi @cgswong,

Sorry this is tripping you up, trying to handle multiple accounts in the backend can be confusing.

The access denied is because when you run `init` and change the backend config, terraform's default behavior is to migrate the state from previous backend to the new backend. So you new configuration may be correct, but you don't probably have the credentials loaded to access the previous state.

This is what the `-reconfigure` flag was added to support, which ignores the previous configuration altogether.

**Comment by izlam** [comment]
Hi,

I found the issue is the order of your providers, if your backend uses a non-default provider, ensure they have the same order.

  module "sample" {
    providers = {
      aws.enterprise = "aws.enterprise"
      aws.default = "aws.default"
    }

**Comment by AJVA2580** [comment]
Hi, 

Are you guys able to share why am I getting this error? 

**please see below -** 

_

Terraform has detected that the configuration specified for the backend
has changed. Terraform will now check for existing state in the backends.


Error inspecting states in the "s3" backend:
    NoSuchBucket: The specified bucket does not exist
	status code: 404, request id: xxxxxxx host id: =xxxxxxxxxxxxxxxxxxxxxx

Prior to changing backends, Terraform inspects the source and destination
states to determine what kind of migration steps need to be taken, if any.
Terraform failed to load the states. The data in both the source and the
destination remain unmodified. Please resolve the above error and try again.

_

I try re creating the bucket in different region but still not working. 

Thanking you in advances for your help

**Comment by shawnmclean** [comment]
`AWS_PROFILE=secondary terraform init` worked for me.

**Comment by chadasapp** [comment]
I'm getting hit with the same bug, sadly.

**Comment by chadasapp** [comment]
From what little I've dug out, the role in the new account tries to hit the old bucket, or vice versa, at some point.

I'd rather not have to setup temporary cross-account-cross-bucket permissions when Terraform has credentials available to the source & destination.

**Comment by christhomas** [comment]
Yeah, I tried to fiddle around a bit just now but it seems to not make any sense why terraform itself would need those permissions since, in both "infrastructure" and "app" projects, I'm using the same credentials to access resources in aws.

**Comment by christhomas** [comment]
This is what I got when I tried with TF_LOG=TRACE

```
2018/11/29 20:59:04 [DEBUG] [aws-sdk-go] <?xml version="1.0" encoding="UTF-8"?>
<Error>
    <Code>AccessDenied</Code>
    <Message>Access Denied</Message>
    <RequestId>CENSORED</RequestId>
    <HostId>CENSORED</HostId>
</Error>
2018/11/29 20:59:04 [DEBUG] [aws-sdk-go] DEBUG: Validate Response s3/ListObjects failed, not retrying, error AccessDenied: Access Denied
	status code: 403, request id: CENSORED, host id: CENSORED
Error inspecting states in the "s3" backend:
    AccessDenied: Access Denied
	status code: 403, request id: CENSORED, host id: CENSORED
```

**Comment by christhomas** [comment]
I've found that my problem was that I am not using a default profile, so I needed to add the profile = "xxxx" to the backend s3 credentials block and then it works :( I should have realised that earlier.

**Comment by arturmartins** [comment]
in my case, I had added the following lines to my `main.tf`:

```
terraform {
  [...]
  backend "s3" {
+    role_arn       = "arn:aws:iam::[REDACTED]:role/[ROLENAME]"
     [...]
  }
}

provider "aws" {
+  assume_role {
+    role_arn = "arn:aws:iam::[REDACTED]:role/[ROLENAME]"
+  }
[...]
}
```

I needed to run
 `terraform init -reconfigure` in order to update `.terraform/terraform.tfstate`
